tests/types/test_time.cc

```diff
@@ -0,0 +1,175 @@
+    EXPECT(Time(2,0,0) == Time("20000"));
+    EXPECT(Time(2,0,0) == Time("020000"));
+
+    EXPECT(Time(20,0,0) == Time("20"));
+    EXPECT(Time(20,0,0) == Time("2000"));
+    EXPECT(Time(20,0,0) == Time("200000"));
+    EXPECT(Time(20,0,0) == Time("20", true));
+    EXPECT_THROWS(Time(30,0,0));
+    EXPECT_THROWS(Time("30"));
+    EXPECT_THROWS(Time("3000"));
+    EXPECT_THROWS(Time("300000"));
+    EXPECT(Time(30,0,0, true) == Time("30", true));
+
+
+    EXPECT(Time(0,3,0) == Time("003"));
+    EXPECT(Time(0,3,0) == Time("0003"));
+    EXPECT(Time(0,3,0) == Time("00300"));
+    EXPECT(Time(0,3,0) == Time("000300"));
+
+    EXPECT(Time(0,0,4) == Time("00004"));
+    EXPECT(Time(0,0,4) == Time("000004"));
+
+    EXPECT(Time(1,23,0) == Time("123"));
+    EXPECT(Time(1,23,0) == Time("0123"));
+    EXPECT(Time(1,23,0) == Time("12300"));
+    EXPECT(Time(1,23,0) == Time("012300"));
+
+    EXPECT(Time(1,23,45) == Time("12345"));
+    EXPECT(Time(1,23,45) == Time("012345"));
+
+    EXPECT_THROWS(Time("25"));
+    EXPECT_THROWS(Time("175"));
+    EXPECT_THROWS(Time("0175"));
+    EXPECT_THROWS(Time("3025"));
+    EXPECT_THROWS(Time("017345"));
+    EXPECT_THROWS(Time("012375"));
+}
+
+CASE("Time format (hh:mm:ss)") {
+
+    EXPECT(Time(0,0,0) == Time("0:0"));
+    EXPECT(Time(0,0,0) == Time("0:00"));
+    EXPECT(Time(0,0,0) == Time("00:0"));
+    EXPECT(Time(0,0,0) == Time("00:00"));
+    EXPECT(Time(0,0,0) == Time("0:00:00"));
+    EXPECT(Time(0,0,0) == Time("00:0:00"));
+    EXPECT(Time(0,0,0) == Time("00:00:0"));
+    EXPECT(Time(0,0,0) == Time("00:00:00"));
+
+    EXPECT(Time(2,0,0) == Time("2:0"));
+    EXPECT(Time(2,0,0) == Time("02:0"));
+    EXPECT(Time(2,0,0) == Time("2:00"));
+    EXPECT(Time(2,0,0) == Time("02:00"));
+    EXPECT(Time(2,0,0) == Time("2:00:00"));
+    EXPECT(Time(2,0,0) == Time("02:00:00"));
+
+    EXPECT(Time(0,3,0) == Time("00:3"));
+    EXPECT(Time(0,3,0) == Time("00:03"));
+    EXPECT(Time(0,3,0) == Time("0:03:00"));
+    EXPECT(Time(0,3,0) == Time("00:3:00"));
+    EXPECT(Time(0,3,0) == Time("00:03:00"));
+
+    EXPECT(Time(0,0,4) == Time("00:00:4"));
+    EXPECT(Time(0,0,4) == Time("00:00:04"));
+
+    EXPECT(Time(1,23,0) == Time("1:23"));
+    EXPECT(Time(1,23,0) == Time("01:23"));
+    EXPECT(Time(1,23,0) == Time("1:23:00"));
+    EXPECT(Time(1,23,0) == Time("01:23:00"));
+
+    EXPECT(Time(1,23,45) == Time("1:23:45"));
+    EXPECT(Time(1,23,45) == Time("01:23:45"));
+
+    EXPECT_THROWS(Time("25"));
+    EXPECT_THROWS(Time("175"));
+    EXPECT_THROWS(Time("0175"));
+    EXPECT_THROWS(Time("3025"));
+    EXPECT_THROWS(Time("017345"));
+    EXPECT_THROWS(Time("012375"));
+}
+
+CASE("Time with unit (__h__m__s)") {
+    EXPECT(Time(2,0,0) == Time("2h"));
+    EXPECT(Time(2,0,0) == Time("0002H"));
+    EXPECT(Time(2,0,0) == Time("120m"));
+    EXPECT(Time(2,0,0) == Time("7200s"));
+
+    EXPECT(Time(0,3,0) == Time("3M"));
+    EXPECT(Time(0,3,0) == Time("180s"));
+
+    EXPECT(Time(1,23,45) == Time("1h23m45s"));
```

Comment:
We ought to have tests that cover such things as 1h2m3s (I.e. where mins and secs are single digit, including following previous terms).